waku/nwaku.go

```diff
@@ -532,12 +533,70 @@ func New(nwakuCfg *WakuConfig, logger *zap.Logger) (*Waku, error) {
 	}, nil
 }
 
+// The event callback sends back the node's ctx to know to which
+// node is the event being emited for. Since we only have a global
+// callback in the go side, We register all the nodes that we create
+// so we can later obtain which instance of `WakuNode` is should
+// be invoked depending on the ctx received
+
+var nodeRegistry map[unsafe.Pointer]*WakuNode
+
+func init() {
+	nodeRegistry = make(map[unsafe.Pointer]*WakuNode)
+}
+
+func registerNode(node *WakuNode) {
+	_, ok := nodeRegistry[node.wakuCtx]
+	if !ok {
+		nodeRegistry[node.wakuCtx] = node
+	}
+}
+
+func unregisterNode(node *WakuNode) {
+	delete(nodeRegistry, node.wakuCtx)
+}
+
 //export globalEventCallback
 func globalEventCallback(callerRet C.int, msg *C.char, len C.size_t, userData unsafe.Pointer) {
-	// This is shared among all Golang instances
-	// TODO-nwaku
-	// self := Waku{wakuCtx: userData}
-	// self.MyEventCallback(callerRet, msg, len)
+	if callerRet == C.RET_OK {
+		eventStr := C.GoStringN(msg, C.int(len))
+		node, ok := nodeRegistry[userData]
+		if ok {
+			node.OnEvent(eventStr)
+		}
+	} else {
+		errMsgField := zap.Skip()
+		if len != 0 {
+			errMsgField = zap.String("error", C.GoStringN(msg, C.int(len)))
+		}
+		log.Error("globalEventCallback retCode not ok", zap.Int("retCode", int(callerRet)), errMsgField)
+	}
+}
+
+type jsonEvent struct {
+	EventType string `json:"eventType"`
+}
+
+func (n *WakuNode) OnEvent(eventStr string) {
+	jsonEvent := jsonEvent{}
+	err := json.Unmarshal([]byte(eventStr), &jsonEvent)
+	if err != nil {
+		n.logger.Error("could not unmarshal nwaku event string", zap.Error(err))
+		return
+	}
+
+	switch jsonEvent.EventType {
+	case "message":
```

Comment:
Shall we give a more meaningful name?
```suggestion
	case "relay-waku-message":
```

Reply:
this is reading the `message` event defined in libwaku here: https://github.com/waku-org/nwaku/blob/ab0c1d4aa0e545c2c74b7e1ea8045823370eee4d/library/events/json_message_event.nim#L86-L87

We can change the name, but we need to change it too in libwaku, in status-go and in every single repo that is using libwaku to receive messages - this will break compatibility everywhere

So I would probably rename the event at a later stage when integrating the bindings everywhere.
Not sure if it's worth breaking all the repos and modifying them now just because of a name change, I think it's better to do it once we are performing integrations across all repos and take advantage that we're already modifying code in all of them